models/surveys.go

```diff
@@ -102,7 +132,75 @@ func NewAPI(db *sql.DB) (*API, error) {
 		GetClassifierTypeSelectorByIDStmt: getClassifierTypeSelectorByIDStmt,
 		GetSurveyRefStmt:                  getSurveyRefStmt,
 		PutSurveyDetailsBySurveyRefStmt:   putSurveyDetailsBySurveyRefStmt,
-	}, nil
+		CreateSurveyStmt:                  createSurvey,
+		GetLegalBasesStmt:                 getLegalBases,
+		GetLegalBasisFromLongNameStmt:     getLegalBasisFromLongName,
+		GetLegalBasisFromRefStmt:          getLegalBasisFromRef }, nil
+}
+
+// PostSurveyDetails endpoint handler - creates a new survey based on JSON in request
+func (api *API) PostSurveyDetails(w http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+
+	var putData Survey
+	err = json.Unmarshal(body, &putData)
+	if err != nil {
+		http.Error(w, "Error unmarshalling JSON", http.StatusBadRequest)
+	}
+
+	shortName := putData.ShortName
+	longName := putData.LongName
+	surveyRef := putData.Reference
+	legalBasisLongName := putData.LegalBasis
+	legalBasisRef := putData.LegalBasisRef
+	var legalBasis LegalBasis;
+	var errorMessage string;
+
+	if legalBasisRef != "" {
+		legalBasis, err = api.getLegalBasisFromRef(legalBasisRef)
+		errorMessage = fmt.Sprintf("Legal basis with reference %v does not exist", legalBasisRef)
+	} else if legalBasisLongName != "" {
+		legalBasis, err = api.getLegalBasisFromLongName(legalBasisLongName)
+		errorMessage = fmt.Sprintf("Legal basis %v does not exist", legalBasisLongName)
+	} else {
+		http.Error(w, "No legal basis specified for survey", http.StatusBadRequest)
+		return
+	}
+
+	if err == sql.ErrNoRows {
+		http.Error(w, errorMessage, http.StatusBadRequest)
+		return
+	} else if err != nil {
+		http.Error(w, fmt.Sprintf("Error getting legal basis - %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	surveyId := uuid.NewV4()
+
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to generate a UUID for new survey - %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	err = api.getSurveyRef(surveyRef)
+
+	if err == sql.ErrNoRows {
+		// Reference is unique - this is good
+		_, err = api.CreateSurveyStmt.Exec(surveyId, surveyRef, shortName, longName, legalBasis.Reference)
+
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Create survey details failed - %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
```

Comment:
this should return http.StatusCreated (201) along with the new resource that has been created

Reply:
Sure - copy and paste bug :-P